rustls/src/conn/unbuffered.rs

```diff
@@ -0,0 +1,449 @@
+            }
+
+            if let Some(msg) = self.core.deframe(None, &mut buffer)? {
+                let mut state =
+                    match mem::replace(&mut self.core.state, Err(Error::HandshakeNotComplete)) {
+                        Ok(state) => state,
+                        Err(e) => {
+                            self.core.state = Err(e.clone());
+                            return Err(e);
+                        }
+                    };
+
+                match self.core.process_msg(msg, state) {
+                    Ok(new) => state = new,
+
+                    Err(e) => {
+                        self.core.state = Err(e.clone());
+                        return Err(e);
+                    }
+                }
+
+                self.core.state = Ok(state);
+            } else if self.wants_write {
+                break (
+                    buffer.pending_discard(),
+                    MustTransmitTlsData { conn: self }.into(),
+                );
+            } else if self
+                .core
+                .common_state
+                .has_received_close_notify
+            {
+                break (buffer.pending_discard(), ConnectionState::ConnectionClosed);
+            } else if self
+                .core
+                .common_state
+                .may_send_application_data
+            {
+                break (
+                    buffer.pending_discard(),
+                    ConnectionState::TrafficTransit(MayEncryptAppData { conn: self }),
+                );
+            } else {
+                break (
+                    buffer.pending_discard(),
+                    ConnectionState::NeedsMoreTlsData { num_bytes: None },
+                );
+            }
+        };
+
+        Ok(UnbufferedStatus { discard, state })
+    }
+}
+
+/// The current status of the `UnbufferedConnection*`
+#[must_use]
+pub struct UnbufferedStatus<'c, 'i, Data> {
+    /// Number of bytes to discard
+    ///
+    /// After the `state` field of this object has been handled, `discard` bytes must be
+    /// removed from the *front* of the `incoming_tls` buffer that was passed to
+    /// the [`UnbufferedConnectionCommon::process_tls_records`] call that returned this object.
+    ///
+    /// This discard operation MUST happen *before*
+    /// [`UnbufferedConnectionCommon::process_tls_records`] is called
+    pub discard: usize,
+
+    /// The current state of the handshake process
+    ///
+    /// This value MUST be handled prior to calling
+    /// [`UnbufferedConnectionCommon::process_tls_records`] again. See the documentation on the
+    /// variants of [`ConnectionState`] for more details.
+    pub state: ConnectionState<'c, 'i, Data>,
+}
+
+/// The state of the [`UnbufferedConnectionCommon`] object
+#[non_exhaustive] // for forwards compatibility; to support caller-side certificate verification
+pub enum ConnectionState<'c, 'i, Data> {
+    /// One, or more, application data record is available
```

Comment:
small typo: "application data record is available" -> "application data records are available"